File: Lab10/q3.py

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>

#define N 5

pthread_mutex_t chopsticks[N];

void* philosopher(void* num) {
    int id = (int)num;

    while (1) {
        printf("Philosopher %d is thinking\n", id);
        sleep(rand() % 3);

        pthread_mutex_lock(&chopsticks[id]);
        pthread_mutex_lock(&chopsticks[(id + 1) % N]);

        printf("Philosopher %d is eating\n", id);
        sleep(2);

        pthread_mutex_unlock(&chopsticks[(id + 1) % N]);
        pthread_mutex_unlock(&chopsticks[id]);

        printf("Philosopher %d finished eating\n", id);
    }
}

int main() {
    pthread_t threads[N];
    int ids[N];

    for (int i = 0; i < N; i++)
        pthread_mutex_init(&chopsticks[i], NULL);

    for (int i = 0; i < N; i++) {
        ids[i] = i;
        pthread_create(&threads[i], NULL, philosopher, &ids[i]);
    }

    for (int i = 0; i < N; i++)
        pthread_join(threads[i], NULL);

    return 0;
}

File: lab revision/q3.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>

int account_balance = 1000;
pthread_mutex_t lock;

void *teller(void *id) {
    for (int i = 0; i < 5; i++) {
        pthread_mutex_lock(&lock);
        int amount = rand() % 200 - 100;
        account_balance += amount;
        printf("Teller %ld updated balance by %d. New balance: %d\n",
               (long)id, amount, account_balance);
        pthread_mutex_unlock(&lock);
        sleep(1);
    }
    pthread_exit(NULL);
}

int main() {
    pthread_t tellers[3];
    pthread_mutex_init(&lock, NULL);

    for (long i = 0; i < 3; i++)
        pthread_create(&tellers[i], NULL, teller, (void *)i);

    for (int i = 0; i < 3; i++)
        pthread_join(tellers[i], NULL);

    pthread_mutex_destroy(&lock);
    printf("Final account balance: %d\n", account_balance);
    return 0;
}
